src/V3Udp.cpp

```diff
@@ -0,0 +1,422 @@
+//         in perfectMatching(). True minimum-weight perfect matching
+//         would produce a better result. How much better is TBD.
+//
+// Code available from: https://verilator.org
+//
+//*************************************************************************
+//
+// Copyright 2003-2024 by Wilson Snyder. This program is free software; you
+// can redistribute it and/or modify it under the terms of either the GNU
+// Lesser General Public License Version 3 or the Perl Artistic License
+// Version 2.0.
+// SPDX-License-Identifier: LGPL-3.0-only OR Artistic-2.0
+//
+//*************************************************************************
+
+#include "V3PchAstNoMT.h"  // VL_MT_DISABLED_CODE_UNIT
+
+#include "V3Udp.h"
+
+#include <map>
+#include <vector>
+
+VL_DEFINE_DEBUG_FUNCTIONS;
+
+//V3Udp add the support for UDP table.
+// For example:
+// table
+//    x 0 1  :   1;
+//    0 ? 1  :   1;
+//    0 1 0  :   0;
+// endtable
+// For every table line, for the input field,
+// two number (mask number and compare number) will
+//  be generated to help make a judegement whether
+// the input field condition is satisfied. For example,
+// for line x 0 1 : 1, mask = 011 cmp = 001, the condition
+// is mask & inputvar == cmp. This pass should be added
+// before V3Inline and V3Tristate.
+
+class UdpVisitor final : public VNVisitor {
+    bool m_inInitial = false;
+    AstVar* m_iFieldVarp = nullptr;  // Input field var of table line.
+    AstVar* m_oFieldVarp = nullptr;  // Output filed var of table line.
+    std::vector<AstVar*> m_inputVars;  // All the input vars in the AstPrimitive.
+    std::vector<AstVar*> m_outputVars;  // All the output vars in the AstPrimitive.
+    AstPrimitive* m_primp = nullptr;
+    AstNodeStmt* m_comboNodeStmtp = nullptr;  // The stmt for all the combinational lines.
+    AstIf* m_lineStmtp = nullptr;  // The stmt for the current combinational lines.
+    bool m_isFirstOutput = false;  // Whether the first IO port is output.
+    AstUdpTableLineVal* m_edgeTrigValp = nullptr;  // The edge trigger value.
+    AstVar* m_trigVarp = nullptr;  // The edge trigger var.
+    std::map<AstVar*, AstAlways*> m_trigAlwaysMapp;  // The always block map for r|f edge.
+    std::map<AstAlways*, AstIf*>
+        m_alwaysIfMapp;  // The map between always and if stmt for sequent line.
+    AstVar* m_newOutVarp = nullptr;  // The output varp.
+    uint32_t m_inputNum = 0;  // The input var number.
+    uint32_t m_outputNum = 0;  // The output var number.
+    AstVarRef* m_outputInitVerfp = nullptr;
+
+    void visit(AstInitial* nodep) {
+        if (m_primp) { m_inInitial = true; }
+        iterateChildren(nodep);
+    }
+    void visit(AstAssign* nodep) {
+        if (m_inInitial) { m_outputInitVerfp = VN_CAST(nodep->lhsp(), VarRef); }
+        iterateChildren(nodep);
+    }
+    void visit(AstPrimitive* nodep) override {
+        m_outputInitVerfp = nullptr;
+        m_primp = nodep;
```

Comment:
Use VL_RESTORED on anything that needs to revert at end of visitor - and remove code that then restored it otherwise.
 
```suggestion
        UASSERT_OBJ(!m_primp, nodep, "Primitives cannot nest");
        VL_RESTORED(m_primp);
        m_outputInitVerfp = nullptr;
        m_primp = nodep;
```